[:has() pseudo-class] Support invalidation for :any-link, :link and :-webkit-any-link

nt1m opened the pull request:
#### 1d40c072bffe76581c75de0c192d379951ca4daa
<pre>
[:has() pseudo-class] Support invalidation for :any-link, :link and :-webkit-any-link
<a href="https://bugs.webkit.org/show_bug.cgi?id=262836">https://bugs.webkit.org/show_bug.cgi?id=262836</a>
<a href="https://rdar.apple.com/116616425">rdar://116616425</a>

Reviewed by Darin Adler.

Move Node::isLink/setIsLink to Element, and make use of Style::PseudoClassChangeInvalidation in the setter.

Add a web platform test as well.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/link-pseudo-class-in-has-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/link-pseudo-class-in-has.html: Added.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setIsLink):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isLink const):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::enclosingLinkEventParentOrSelf):
* Source/WebCore/dom/Node.h:
(WebCore::Node::isLink const): Deleted.
(WebCore::Node::setIsLink): Deleted.
* Source/WebCore/editing/Editing.cpp:
(WebCore::enclosingAnchorElement):
* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::attributeChanged):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::disallowTelephoneNumberParsing):
* Source/WebCore/mathml/MathMLElement.cpp:
(WebCore::MathMLElement::attributeChanged):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::addPDFURLRect const):
* Source/WebCore/svg/SVGAElement.cpp:
(WebCore::SVGAElement::svgAttributeChanged):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::sendTapHighlightForNodeIfNecessary):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode _linkElement]):

Canonical link: <a href="https://commits.webkit.org/271919@main">https://commits.webkit.org/271919@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/384cf40f9d29e0a9e9a975b07b3260f889a74206

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30064 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8728 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31378 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32565 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27180 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30733 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10924 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5975 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27205 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30362 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7318 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25639 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6247 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6405 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/26732 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33901 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27413 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27163 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32585 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6348 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4517 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30384 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8097 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/26532 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/7119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7101 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3876 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6876 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2342,6 +2342,16 @@ URL Element::absoluteLinkURL() const
 
+void Element::setIsLink(bool flag)
+{
+    Style::PseudoClassChangeInvalidation styleInvalidation(*this, {
```
Now that we are doing invalidation here, I suggest an early return for the case where the flag is not changing to avoid unnecessary work.

nt1m replied:
Will add one!

darinadler commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -755,14 +755,11 @@ void RenderObject::addPDFURLRect(const PaintInfo& paintInfo, const LayoutPoint&
-    }();
-    if (!element)
+
+    RefPtr element = dynamicDowncast<Element>(this->node());
```
Just `node()` is fine now. The `this->` was needed in the old code because the name `node` was also a local variable.

annevk commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -217,11 +217,14 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
+        if (!element)
+            continue;
+        bool matchedButton = is<HTMLButtonElement>(element);
+        bool matchedLabel = isLabelable && is<HTMLLabelElement>(element);
```
I think you can elide null checks here?

nt1m replied:
Fixed

nt1m committed:
[:has() pseudo-class] Support invalidation for :any-link, :link and :-webkit-any-link
https://bugs.webkit.org/show_bug.cgi?id=262836
rdar://116616425

Reviewed by Darin Adler.

Move Node::isLink/setIsLink to Element, and make use of Style::PseudoClassChangeInvalidation in the setter.

Add a web platform test as well.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/link-pseudo-class-in-has-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/link-pseudo-class-in-has.html: Added.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setIsLink):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isLink const):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::enclosingLinkEventParentOrSelf):
* Source/WebCore/dom/Node.h:
(WebCore::Node::isLink const): Deleted.
(WebCore::Node::setIsLink): Deleted.
* Source/WebCore/editing/Editing.cpp:
(WebCore::enclosingAnchorElement):
* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::attributeChanged):
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::disallowTelephoneNumberParsing):
* Source/WebCore/mathml/MathMLElement.cpp:
(WebCore::MathMLElement::attributeChanged):
* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::addPDFURLRect const):
* Source/WebCore/svg/SVGAElement.cpp:
(WebCore::SVGAElement::svgAttributeChanged):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::sendTapHighlightForNodeIfNecessary):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode _linkElement]):

Canonical link: https://commits.webkit.org/271919@main
